stan/math/prim/mat/fun/value_of_rec.hpp

```diff
@@ -22,12 +22,33 @@ namespace stan {
     template <typename T, int R, int C>
     inline Eigen::Matrix<double, R, C>
     value_of_rec(const Eigen::Matrix<T, R, C>& M) {
-      using stan::math::value_of_rec;
-      Eigen::Matrix<double, R, C> Md(M.rows(), M.cols());
-      for (int j = 0; j < M.cols(); j++)
-        for (int i = 0; i < M.rows(); i++)
-          Md(i, j) = value_of_rec(M(i, j));
-      return Md;
+      Eigen::Matrix<double, R, C>
+        result(M.rows(), M.cols());
+
+      double * data_r = result.data();
+      const T * data_m = M.data();
+
+      int S = M.size();
+      for (int i=0; i < S; i++)
+        data_r[i] = value_of_rec(data_m[i]);
+      return result;
+    }
```

Comment:
> I understand not doing the double indexing --- that's slow. But I 
> think the bit after the declaration/construction of result, we should 
> just do the following unless you can demonstrate that what you did 
> with all the temporaries is faster.
>
> |for (int i = 0; i < M.size(); ++i) result(i) = value_of(M(i)); |
>

The following code https://gist.github.com/randommm/4e61fabe8c5438daeebf

runs a kahan sum usingfour different methods:

(warning: requires more than 3GB to run)

(1)
for (size_t i = 0, nRows = xs.rows(), nCols = xs.cols(); i < nCols; ++i)
   for (size_t j = 0; j < nRows; ++j)
//do something with matrix(i, j)

(2)
for (size_t i = 0, nRows = xs.rows(), nCols = xs.cols(); i < nRows; ++i)
   for (size_t j = 0; j < nCols; ++j)
//do something with matrix(i, j)

(3)
for (size_t i = 0, size = xs.size(); i < size; i++)
   //do something with matrix(i)

(4)
T * matrix_data = matrix.data();
for (size_t i = 0, size = xs.size(); i < size; i++)
//do something with matrix_data[i]


Here's the output of how much time each of them took:

$ g++ main.cpp `pkg-config eigen3 --cflags --libs`
$ ./a.out
start
25
49
25
7

The first run mat(i, j)

$ g++ main.cpp -DEIGEN_NO_DEBUG `pkg-config eigen3 --cflags --libs`
$ ./a.out
start
13
35
12
7